### DISCUSSION TITLE
ClojureCLR build tooling

### DISCUSSION THREAD

**Discussion by dmiller** [dmiller/clojure-clr-next#5] [Ideas]
With the advent of `cljr`, ClojureCLR now has some of the goodness of the CLI/deps.edn tooling available to Clojure(JVM).

Some functionality is still missing, centered around tool definition and invocation and project preparation.  In the [Prep libs](https://clojure.org/reference/clojure_cli#deps_prep): section of the CLI guide, we find: 

> Source libs with Clojure source can immediately be added to the classpath of a project using it. However, some source libs require some preparation before they can be added, for example due to needing Java compilation, or copying / replacing resource files, etc. 

The library [tools.build](https://clojure.org/guides/tools_build) provides functionality for building Clojure projects.  It is very JVM-specific, so I have not bothered to port it.  We need our own.

Questions to contemplate:

- What would be in a `tools.build` for ClojureCLR?
- How do we add `:nuget` coordinates to `cljr`?  Being able to retrieve packages, dealing with versioning, etc., is straightforward enough, but there are other issues:
    - How to deal with multiple frameworks/runtimes?
    - How best to package pure Clojure code projects?  (The current method is clunky. Is there a better way?)
- How do we specify the version of ClojureCLR in a `deps.edn` file?  (This can be done for Clojure.  However, the differences between Java classpaths and what we have to do with assemblies in ClojureCLR prevents a similar approach -- unless perhaps we solve all the other problems here.)
- Are there better ways for ClojureCLR to reference assemblies?

Let's go!

(Note:  this discussion is hosted on the clojure-clr-next repo because I own it and can do things such as set up discussions.  However, this is really a discussion about ClojureCLR as it is now.)

**Comment by brandoncorrea** [comment]
> What would be in a `tools.build` for ClojureCLR?

`tools.build` seems to capture discrete steps of the Java build process. It would be neat if the CLR version did the same.

Maybe it would include some of these steps:
- Create `.sln` and `.csproj` files (probably `.fsproj` and `.vbproj` as well)
- Compile .NET code to something like `/bin` or `/out` (or any user-specified output directory)
- Packaging...
  - The JVM has jars. Does .NET have such a thing? In my experience, it has always been a hundred files in a directory with a main.exe nested somewhere within.

It would be nice to have some sort of handling for users with mixed sources (`clj` + `fs` + `cs`). How does the JVM version handle mixed languages?

**Reply by dmiller** [reply]
A lot of functionality can be offloaded to the `dotnet` command.
Re packaging:  Of course, assemblies serve part of the purpose of packaging.  But beyond, the closest thing I know of is nuget packages.
Re mixed languages: Everything compiles to bytecode in .class files and ends up in jars, uberjars, or just the files themselves sitting on the classpath.  It's very uniform.  And relatively transparent, if you don't mind perusing through jar files to find .class files.  In .NET-land, of course we can compile to IL, but identity and access is assembly-based, and that lumpiness is what we need to work around.

**Comment by brandoncorrea** [comment]
> Are there better ways for ClojureCLR to reference assemblies?

How does .NET do this? Does it not make sense to add it to `:deps` and maybe extend the `:mvn/repos` section?

**Reply by dmiller** [reply]
The two requirements are to find ways to specify in `deps.edn` and then to make sure ClojureCLR itself has the mechanisms to handle it.
One could reference an assembly via a `:local` for something on the filesystem, or a `:nuget` coordinate (when implemented).  But ClojureCLR does not have an automatic way to pull in assemblies based on that, so that would have to be added.

**Reply by brandoncorrea** [reply]
I think that makes sense then, `:nuget` for deployed packages and `:local` for assemblies & projects. That's how `.*proj` references seem to work (to my knowledge).

**Comment by brandoncorrea** [comment]
> How do we specify the version of ClojureCLR in a deps.edn file?

Java has `{:jvm-opts ["--release 11"]}`. Why not something like `{:clr-opts ["..."]}`? Unless these are too different to be as closely related as they look.

**Reply by dmiller** [reply]
Rather than the version of the CLR, I was thinking of the version of ClojureCLR itself.  On the Clojure(JVM) side, you can find examples such as:

```Clojure
{:paths ["src/main/clojure"]
 :deps {org.clojure/clojure {:mvn/version "1.9.0"}}}
```
That works to bring down the 1.9 version of Clojure.  Once that is on the classpath, that's what you are going to pick up when you run.
You run `java`, specify that `clojure.main` or whatever is to be started, and it is found on the classpath.

A bit more problematic on the CLR side.  Currently the ClojureCLR runtime installs as a tool.  Likely we could install as a local tool to get versioning, but that definitely would not be part of the standard `:deps` processing.   Or we bring down the assembly using a `:nuget` coordinate, and use `dotnet run` as our startup rather than invoking the tool.  Or something else.  We don't have a direct equivalent of `java clojure.main` (or whatever the proper invocation is.

**Reply by brandoncorrea** [reply]
It seems as though the JVM side brings down the correct Clojure version from [Maven](https://mvnrepository.com/artifact/org.clojure/clojure). I think it would make sense that ClojureCLR do the same with [Nuget](https://www.nuget.org/packages/Clojure/1.12.0-alpha9).

If I'm not mistaken, the Clojure JVM `clj` command is a tool, but different from the Clojure version specified in `deps.edn`, allowing the project to run a different version of Clojure from that of the `clj` command.

This would _probably_ affect how the `cljr` command works though, and maybe some other existing tools? Any reason not to take the same approach as JVM?

In this scenario, I wonder if `cljr` could be so decoupled that it _could_ be written on the JVM, if so desired (though I imagine .NET is our friend here).

**Comment by brandoncorrea** [comment]
> How do we add :nuget coordinates to cljr? Being able to retrieve packages, dealing with versioning, etc., is straightforward enough, but there are other issues:

I believe `:mvn/version` fetches packages from either Maven or Clojars (and anywhere else you specify in `:mvn/repos`). Would CLR's version (`:nuget/version` & `:nuget/repos` maybe?) do something similar by fetching from either NuGet or Clojars?

> How to deal with multiple frameworks/runtimes?

Whose responsibility is this? Clojure JVM has supported all versions of Java from 1.8 up until recently. Clojure CLR I think should certainly do something similar, but as far as packages go, wouldn't that be up to the maintainer who they want to support?

> How best to package pure Clojure code projects? (The current method is clunky. Is there a better way?)

Clojars? Or do we want to deploy these to nuget as well?

**Reply by brandoncorrea** [reply]
On `:nuget/repos`, I'm pretty sure private nuget repos are a thing. So it may be worth considering. Maybe _very_ little need though, assuming packages can be pulled from private git repos (with proper SSH keys).

**Reply by dmiller** [reply]
I'm thinking NuGet packages to bring down non-Clojure dependencies -- the library written in C# that you need for your project.
The problem is how to do loading of those assemblies.  I found code to do pulls, manage version numbering, etc., from Nuget.  I could port that for use in the CLI / deps.edn.  Bringing down the bytes is the easy part.  I'm not sure how to deal with platform versioning, other files in the download, etc.

**Comment by Dangercoder** [comment]
It would be nice to be able to test, run, deploy and run a nrepl via `.csproj` and `msbuild`. No `deps.edn`. What hurts adoption (from my .NET friends who wants to try ClojureCLR) is that they have to install the "clojure tools" in order to try it out. 

Example on how to do this here: https://github.com/Dangercoder/Clojure.MSBuild

**Reply by jamesdavidson** [reply]
In the spirit of making Clojure more approachable to .NET friends I also went down the path of using `.csproj` as the entrypoint but found a way to include libraries specified in `deps.edn` by way of a prebuild step that bundles the sources into a DLL.

I'm intrigued by `dotnet test` integration part of your project but I must confess I haven't actually cloned the repo and tried it out yet. I fell for the temptation to write code rather than to read it!

Anyway here is a link to my example project repo extracted from some work I was doing with Parquet.Net (from NuGet) and data.csv (from GitHub): https://github.com/jamesdavidson/prebuild-example

<img width="1922" height="1048" alt="prebuild-example-2025-12-01" src="https://github.com/user-attachments/assets/93cc759a-f096-4663-8fbe-61a86ea5f25f" />

I'm curious to know how other people go about mixing Clojure and C# dependencies in their projects and what you think of my approach.

**Comment by jamesdavidson** [comment]
I recently ported clojure.data.int-map and I think it's useful to share here as an example of how a library that mixes C# and Clojure can be built and packaged.

https://github.com/jamesdavidson/clr.data.int-map

Now I suppose the question is how can we extend `cljr` so that it can pull this library from NuGet (if it were published).

**Comment by dmiller** [comment]
It might take me a day or two to digest it.

You stated previously "I'm curious to know how other people go about mixing
Clojure and C# dependencies in their projects..."
Me, too.
I crave feedback.

I do think the next priority for ClojureCLR development is dealing with
these issues.
